tests/test_graphnet_backend.py

```diff
@@ -0,0 +1,423 @@
+from datetime import date, datetime
+
+import pytest
+from lib.graphnet_schema import (  # ClinicalEventsSnomed,; Hospitalizations,; PatientAddress,; hospitalization,
+    ClinicalEvents,
+    CovidTestResults,
+    Patients,
+    PracticeRegistrations,
+    patient,
+    patient_address,
+    registration,
+)
+from lib.util import extract
+
+from cohortextractor import table
+#  from cohortextractor import codelist, table
+from cohortextractor.backends.graphnet import GraphnetBackend
+
+
+@pytest.mark.integration
+def test_basic_events_and_registration(database, setup_backend_database):
+    setup_backend_database(
+        Patients(Patient_ID=1),
+        PracticeRegistrations(Patient_ID=1),
+        ClinicalEvents(Patient_ID=1, CTV3Code="Code1"),
+        backend="graphnet",
+    )
+
+    class Cohort:
+        code = table("clinical_events").first_by("patient_id").get("code")
+
+    assert extract(Cohort, GraphnetBackend, database) == [
+        dict(patient_id=1, code="Code1")
+    ]
+
+
+@pytest.mark.integration
+def test_registration_dates(database, setup_backend_database):
+    setup_backend_database(
+        Patients(Patient_ID=1),
+        PracticeRegistrations(
+            Patient_ID=1, StartDate="2001-01-01", EndDate="2012-12-12"
+        ),
+        PracticeRegistrations(Patient_ID=1, StartDate="2013-01-01"),
+        backend="graphnet",
+    )
+
+    class Cohort:
+        _registrations = table("practice_registrations").first_by("patient_id")
```

Comment:
Could we also add a test that uses a Cohort like
```
    class Cohort:
        _registrations = table("practice_registrations").date_in_range("2014-01-01")
        arrived = _registrations.first_by("patient_id").get("date_start")
        left = _registrations.first_by("patient_id").get("date_end")
```
?

The `date_in_range` bit should select only registrations where a patient was registered on 2014-01-01 (i.e. the entry with a null end date).

It _should_ return
`dict(patient_id=1, arrived=datetime(2013, 1, 1), left=None)`

but I am fairly certain it'll fail and it won't select the entry with no end date (and that's something we need to fix in the query language/engine).

Reply:
If it fails as I think it will, we can either leave the test for it commented out and make a note of it in the new ticket to fix it, or mark it as an xfail (with `@pytest.mark.xfail`) for now